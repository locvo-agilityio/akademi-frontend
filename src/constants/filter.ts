export const FILTER_OPTIONS = [
  {
    value: 'desc',
    label: 'Descending',
  },
  {
    value: 'asc',
    label: 'Ascending',
  },
];
