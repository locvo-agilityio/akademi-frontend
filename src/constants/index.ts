export * from './sidebar';
export * from './message';
export * from './router';
export * from './filter';
export * from './httpsRequest';
export * from './api';
export * from './query';
export * from './pagination';
export * from './teacher';
export * from './file';
export * from './form';
export * from './regex';
export * from './status';
export * from './student';
