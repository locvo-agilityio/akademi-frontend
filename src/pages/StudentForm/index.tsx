import { lazy, Suspense } from 'react';
import { useParams } from 'react-router-dom';

// Components
import { Fallback } from '@/components';

// Hooks
import { useGetStudent } from '@/hooks';

const StudentFormUI = lazy(() => import('@/ui/StudentForm/'));

const StudentForm = () => {
  const { id } = useParams();
  const { student, isStudentLoading } = useGetStudent(id ?? '');

  return (
    <Suspense fallback={<Fallback />}>
      {!isStudentLoading && <StudentFormUI defaultValues={student?.data} />}
    </Suspense>
  );
};

export default StudentForm;
