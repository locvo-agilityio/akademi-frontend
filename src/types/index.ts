export * from './table';
export * from './router';
export * from './filter';
export * from './httpsRequest';
export * from './queryKey';
export * from './student';
export * from './pagination';
export * from './teacher';
export * from './schedule';
export * from './file';
export * from './field';
