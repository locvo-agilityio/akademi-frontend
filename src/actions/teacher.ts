'use server';

import { QueryFunctionContext } from '@tanstack/react-query';

// Constants
import {
  API_PATHS,
  HTTP_METHOD,
  QUERY,
  teacherKeys,
  teachersQueryKeys,
} from '@/constants';

// Services
import { apiRequest } from '@/services';

// Types
import { ITeacher, ITeachers } from '@/types';

// Mocks
import { MOCK_TEACHER_SCHEDULES } from '@/__mocks__';

export const getTeachers = async ({
  queryKey: [{ page, filter }],
}: QueryFunctionContext<ReturnType<(typeof teachersQueryKeys)['list']>>) => {
  const search = filter?.search
    ? `&filters[email][$contains]=${filter.search}`
    : '';

  const sort = filter?.sort ? `&sort=createdAt:${filter.sort}` : '';

  const url = `${API_PATHS.TEACHERS}?${QUERY.TEACHERS(page)}${search}${sort}`;

  return await apiRequest<ITeachers>(HTTP_METHOD.GET, url);
};

export const getTeacher = async ({
  queryKey: [{ id }],
}: QueryFunctionContext<ReturnType<(typeof teachersQueryKeys)['detail']>>) => {
  const url = `${API_PATHS.TEACHERS}/${id}`;

  return await apiRequest<{ data: ITeacher }>(HTTP_METHOD.GET, url);
};

export const addTeacher = async (_: unknown, formData: FormData) => {
  const payload: ITeacher = {
    ...(Object.fromEntries(
      teacherKeys.map((key) => [key, formData.get(key) as string]),
    ) as unknown as ITeacher),
    schedule: MOCK_TEACHER_SCHEDULES,
  };

  return await apiRequest(HTTP_METHOD.POST, API_PATHS.TEACHERS, {
    data: payload,
  });
};

export const editTeacher = async (_: unknown, formData: FormData) => {
  const documentId = formData.get('documentId');
  const url = `${API_PATHS.TEACHERS}/${documentId}`;

  const payload: ITeacher = {
    ...(Object.fromEntries(
      teacherKeys.map((key) => [key, formData.get(key) as string]),
    ) as unknown as ITeacher),
    schedule: MOCK_TEACHER_SCHEDULES,
  };

  return await apiRequest(HTTP_METHOD.PUT, url, {
    data: payload,
  });
};

export const deleteTeacher = async ({ id }: { id: string }) => {
  const url = `${API_PATHS.TEACHERS}/${id}`;

  return await apiRequest(HTTP_METHOD.DELETE, url);
};
