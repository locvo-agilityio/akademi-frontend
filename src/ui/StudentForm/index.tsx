import {
  Button,
  Flex,
  FormControl,
  FormLabel,
  HStack,
  Radio,
  RadioGroup,
  Stack,
  Text,
  useToast,
  VStack,
} from '@chakra-ui/react';
import { Controller, useForm } from 'react-hook-form';
import { useQueryClient } from '@tanstack/react-query';
import { Link, useNavigate } from 'react-router-dom';
import { useActionState, useCallback, useEffect, useTransition } from 'react';

// Components
import {
  CustomToast,
  InputField,
  TextareaField,
  UploadImage,
} from '@/components';

// Constants
import {
  CLASS_NAME,
  CLASS_NAME_OPTIONS,
  ERROR_MESSAGES,
  PUBLIC_ROUTERS,
  studentsQueryKeys,
  SUCCESS_MESSAGES,
  TOAST_STATUS,
  VALIDATION_RULES,
} from '@/constants';

// Types
import { IStudent } from '@/types';

// Utils
import {
  ADDITIONAL_PARENT_STUDENT_INFO,
  BASIC_PARENT_STUDENT_INFO,
  BASIC_STUDENT_INFO,
} from '@/utils';

// Actions
import { addStudent, editStudent } from '@/actions';

interface IStudentFormProps {
  defaultValues?: IStudent;
}

const StudentForm = ({ defaultValues }: IStudentFormProps) => {
  const toast = useToast();
  const queryClient = useQueryClient();
  const navigate = useNavigate();
  const [isPending, startTransition] = useTransition();
  const isEdit = !!defaultValues?.documentId;

  const actionForm = isEdit ? editStudent : addStudent;

  const [state, formAction] = useActionState(actionForm, undefined);
  const {
    firstName = '',
    lastName = '',
    parentName = '',
    phone = '',
    email = '',
    address = '',
    city = '',
    grade = CLASS_NAME.VII_A,
    avatar = '',
    birthday = '',
    parent = { firstName: '', lastName: '', email: '', phone: '', address: '' },
    payment = 'unpaid',
    documentId = '',
  } = defaultValues || {};

  const initialValues: IStudent = {
    firstName: firstName,
    lastName: lastName,
    parentName: parentName,
    phone: phone,
    email: email,
    address: address,
    city: city,
    grade: grade,
    avatar: avatar,
    birthday: birthday,
    parent: parent,
    payment: payment,
    documentId: documentId,
  };

  const {
    control,
    setValue,
    getValues,
    formState: { isValid, isDirty },
  } = useForm<IStudent>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
    defaultValues: initialValues,
  });

  const handleFileChange = useCallback(
    (file: string) => {
      setValue('avatar', file, { shouldDirty: true });
    },
    [setValue],
  );

  const actionSubmit = useCallback(
    async (formData: FormData) => {
      try {
        startTransition(() => {
          formData.append('avatar', getValues('avatar'));

          if (isEdit)
            formData.append('documentId', defaultValues?.documentId as string);

          formAction(formData);

          queryClient.invalidateQueries({
            queryKey: studentsQueryKeys.lists(),
          });

          toast({
            render: () => (
              <CustomToast
                status={TOAST_STATUS.SUCCESS}
                message={
                  isEdit
                    ? SUCCESS_MESSAGES.EDIT_STUDENT_SUCCESS
                    : SUCCESS_MESSAGES.ADD_STUDENT_SUCCESS
                }
              />
            ),
          });
        });
      } catch (err) {
        toast({
          render: () => (
            <CustomToast
              status={TOAST_STATUS.ERROR}
              message={
                isEdit
                  ? ERROR_MESSAGES.EDIT_STUDENT_FAILED
                  : ERROR_MESSAGES.ADD_STUDENT_FAILED
              }
            />
          ),
        });
      }
    },
    [defaultValues?.documentId, getValues, isEdit, queryClient, toast],
  );

  useEffect(() => {
    if (state) {
      queryClient.setQueryData(
        studentsQueryKeys.detail(state.data.documentId),
        state,
      );

      navigate(PUBLIC_ROUTERS.STUDENTS);
    }
  }, [navigate, queryClient, state]);

  return (
    <VStack as="form" alignItems="flex-end" w="full" action={actionSubmit}>
      <VStack mt={5} w="full" bgColor="white" borderRadius="2xl">
        <Text
          as="h3"
          w="full"
          px={10}
          h="60px"
          borderTopRadius="2xl"
          bgColor="primary"
          color="white"
          size="lg"
          fontWeight="bold"
          display="flex"
          alignItems="center"
        >
          Student Details
        </Text>

        <HStack w="full" alignItems="flex-start" p={8} gap={10}>
          <Controller
            key="avatar"
            control={control}
            name="avatar"
            render={() => (
              <UploadImage imageUrl={avatar} onFileChange={handleFileChange} />
            )}
          />

          <Flex w="full" gap={7} direction="column" justifyContent="flex-start">
            <Controller
              key="firstName"
              control={control}
              rules={VALIDATION_RULES.FIRST_NAME}
              name="firstName"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  label="First Name *"
                  placeholder="Please enter your first name"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  aria-label="first name"
                />
              )}
            />

            <Flex
              direction={{ base: 'column', xl: 'row' }}
              gap={6}
              alignItems="flex-end"
            >
              <Controller
                key="birthday"
                control={control}
                rules={VALIDATION_RULES.BIRTH_DATE}
                name="birthday"
                render={({ field, fieldState: { error } }) => (
                  <InputField
                    label="Date & Place of Birth *"
                    type="date"
                    max="2009-12-31"
                    placeholder="Please enter your birthday"
                    {...field}
                    isError={!!error}
                    errorMessages={error?.message}
                    aria-label="birthday"
                  />
                )}
              />

              <Controller
                key="city"
                control={control}
                rules={VALIDATION_RULES.CITY}
                name="city"
                render={({ field, fieldState: { error } }) => (
                  <InputField
                    placeholder="Please enter your city"
                    {...field}
                    isError={!!error}
                    errorMessages={error?.message}
                    aria-label="city"
                  />
                )}
              />
            </Flex>

            <Controller
              key="email"
              control={control}
              rules={VALIDATION_RULES.EMAIL}
              name="email"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  label="Email *"
                  type="email"
                  placeholder="Please enter your email"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  aria-label="email"
                />
              )}
            />

            <Controller
              key="address"
              control={control}
              rules={VALIDATION_RULES.ADDRESS}
              name="address"
              render={({ field, fieldState: { error } }) => (
                <TextareaField
                  label="Address"
                  placeholder="Please enter your address"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  aria-label="address"
                />
              )}
            />
          </Flex>

          <Flex w="full" gap={7} direction="column" justifyContent="flex-start">
            {BASIC_STUDENT_INFO.map(
              ({ name, label, placeholder, type, rules, ariaLabel }) => (
                <Controller
                  key={name}
                  control={control}
                  rules={rules}
                  name={name}
                  render={({ field, fieldState: { error } }) => (
                    <InputField
                      label={label}
                      placeholder={placeholder}
                      type={type}
                      {...field}
                      isError={!!error}
                      errorMessages={error?.message}
                      aria-label={ariaLabel}
                    />
                  )}
                />
              ),
            )}
            <Controller
              key="grade"
              control={control}
              name="grade"
              render={({ field: { onChange, value } }) => (
                <FormControl>
                  <FormLabel
                    fontSize="md"
                    fontWeight="semibold"
                    marginInlineEnd={0}
                    minW="max-content"
                    color="darkBlue"
                  >
                    Class *
                  </FormLabel>

                  <RadioGroup
                    onChange={onChange}
                    value={value}
                    size="sm"
                    display="flex"
                    gap={10}
                  >
                    {CLASS_NAME_OPTIONS.map((option) => (
                      <Radio
                        key={option.value}
                        name={option.name}
                        value={option.value}
                        colorScheme={option.colorScheme}
                      >
                        <Text
                          p={2}
                          fontSize="xs"
                          color="white"
                          alignItems="center"
                          bgColor={option.bgColor}
                          justifyContent="center"
                          borderRadius="full"
                        >
                          {option.text}
                        </Text>
                      </Radio>
                    ))}
                  </RadioGroup>
                </FormControl>
              )}
            />
          </Flex>
        </HStack>
      </VStack>

      <VStack mt={5} w="full" bgColor="white" borderRadius="2xl">
        <Text
          as="h3"
          w="full"
          px={10}
          h="60px"
          borderTopRadius="2xl"
          bgColor="primary"
          color="white"
          size="lg"
          fontWeight="bold"
          display="flex"
          alignItems="center"
        >
          Parent Details
        </Text>

        <HStack w="full" alignItems="flex-start" p={8} gap={10}>
          <Flex w="full" gap={7} direction="column" justifyContent="flex-start">
            {BASIC_PARENT_STUDENT_INFO.map(
              ({ name, label, placeholder, type, rules, ariaLabel }) => (
                <Controller
                  key={name}
                  control={control}
                  rules={rules}
                  name={name}
                  render={({ field, fieldState: { error } }) => (
                    <InputField
                      label={label}
                      placeholder={placeholder}
                      type={type}
                      {...field}
                      isError={!!error}
                      errorMessages={error?.message}
                      aria-label={ariaLabel}
                    />
                  )}
                />
              ),
            )}

            <Controller
              key="address"
              control={control}
              rules={VALIDATION_RULES.ADDRESS}
              name="parent.address"
              render={({ field, fieldState: { error } }) => (
                <TextareaField
                  label="Address"
                  placeholder="Please enter your address"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  aria-label="address"
                />
              )}
            />
          </Flex>

          <Flex w="full" gap={7} direction="column" justifyContent="flex-start">
            {ADDITIONAL_PARENT_STUDENT_INFO.map(
              ({ name, label, placeholder, type, rules, ariaLabel }) => (
                <Controller
                  key={name}
                  control={control}
                  rules={rules}
                  name={name}
                  render={({ field, fieldState: { error } }) => (
                    <InputField
                      label={label}
                      placeholder={placeholder}
                      type={type}
                      {...field}
                      isError={!!error}
                      errorMessages={error?.message}
                      aria-label={ariaLabel}
                    />
                  )}
                />
              ),
            )}

            <Controller
              key="payment"
              control={control}
              name="payment"
              render={({ field: { onChange, value } }) => (
                <FormControl>
                  <FormLabel
                    fontSize="md"
                    fontWeight="semibold"
                    marginInlineEnd={0}
                    minW="max-content"
                    color="darkBlue"
                  >
                    Payments *
                  </FormLabel>

                  <RadioGroup onChange={onChange} value={value} size="sm">
                    <Stack direction="row">
                      <Radio
                        name="payment"
                        value="cash"
                        colorScheme="purple"
                        defaultChecked
                      >
                        Cash
                      </Radio>
                      <Radio name="payment" value="debit" colorScheme="purple">
                        Debit
                      </Radio>
                      <Radio name="payment" value="unpaid" colorScheme="purple">
                        Unpaid
                      </Radio>
                    </Stack>
                  </RadioGroup>
                </FormControl>
              )}
            />
          </Flex>
        </HStack>
      </VStack>

      <HStack mt={4} gap={3}>
        <Button
          as={Link}
          to={PUBLIC_ROUTERS.STUDENTS}
          w="full"
          p={7}
          variant="secondary"
          borderRadius="full"
        >
          Cancel
        </Button>

        <Button
          w="full"
          p={7}
          variant="primary"
          borderRadius="full"
          type="submit"
          isLoading={isPending}
          isDisabled={isPending || !isValid || !isDirty}
        >
          Submit
        </Button>
      </HStack>
    </VStack>
  );
};

export default StudentForm;
