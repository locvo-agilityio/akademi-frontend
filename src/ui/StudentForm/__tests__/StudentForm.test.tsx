import { ReactNode } from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { MemoryRouter } from 'react-router-dom';
import { ChakraProvider } from '@chakra-ui/react';
import userEvent from '@testing-library/user-event';

// Actions
import { addStudent, editStudent } from '@/actions';

// Constants
import { CLASS_NAME } from '@/constants';

// Types
import { IStudent } from '@/types';

// Components
import StudentForm from '..';

// Mock actions
jest.mock('@/actions', () => ({
  addStudent: jest.fn(),
  editStudent: jest.fn(),
}));

const queryClient = new QueryClient();

const renderWithProviders = (component: ReactNode) =>
  render(
    <QueryClientProvider client={queryClient}>
      <MemoryRouter
        future={{ v7_startTransition: true, v7_relativeSplatPath: true }}
      >
        <ChakraProvider>{component}</ChakraProvider>
      </MemoryRouter>
    </QueryClientProvider>,
  );

const defaultValues: IStudent = {
  firstName: 'Johnny ',
  lastName: 'Ahmad',
  birthday: '1999-03-25',
  city: 'Jakarta',
  parentName: 'John Doe',
  email: 'john.doe@example.com',
  phone: '2345678901',
  address: 'Jakarta',
  grade: CLASS_NAME.VII_A,
  avatar: '',
  parent: {
    firstName: 'John',
    lastName: 'Doe',
    email: 'john.doe@example.com',
    phone: '2345678901',
    address: 'Jakarta',
  },
  payment: 'unpaid',
};

describe('StudentForm', () => {
  const defaultProps = {
    defaultValues: defaultValues,
  };

  it('renders snapshot with default values', () => {
    const { container } = renderWithProviders(
      <StudentForm {...defaultProps} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders snapshot with empty values', () => {
    const { getAllByPlaceholderText } = renderWithProviders(<StudentForm />);

    const firstNameInput = getAllByPlaceholderText(
      'Please enter your first name',
    )[0];

    expect(firstNameInput).toBeInTheDocument();
  });

  it('validates required fields on blur', async () => {
    const { getAllByLabelText, getByText } = renderWithProviders(
      <StudentForm {...defaultProps} />,
    );

    const firstNameInput = getAllByLabelText('First Name *')[0];
    fireEvent.change(firstNameInput, { target: { value: '' } });
    fireEvent.blur(firstNameInput);

    await waitFor(() => {
      expect(getByText(/first name is required/i)).toBeInTheDocument();
    });
  });

  it('submits the form and calls addStudent for new students', async () => {
    const {
      getAllByLabelText,
      getByRole,
      getByLabelText,
      getByPlaceholderText,
    } = renderWithProviders(<StudentForm />);

    const firstNameInput = getAllByLabelText('First Name *')[0];
    fireEvent.change(firstNameInput, { target: { value: 'Jane' } });

    const lastNameInput = getAllByLabelText('Last Name *')[0];
    fireEvent.change(lastNameInput, { target: { value: 'Smith' } });

    const birthdayInput = getByLabelText('Date & Place of Birth *');
    fireEvent.change(birthdayInput, { target: { value: '2000-01-01' } });

    const cityInput = getByPlaceholderText('Please enter your city');
    fireEvent.change(cityInput, { target: { value: 'New York' } });

    const parentNameInput = getByLabelText('Parent Name *');
    fireEvent.change(parentNameInput, { target: { value: 'Jane Smith' } });

    const emailInput = getAllByLabelText('Email *')[0];
    fireEvent.change(emailInput, {
      target: { value: 'jane.smith@example.com' },
    });
    const phoneInput = getAllByLabelText('Phone *')[0];
    fireEvent.change(phoneInput, { target: { value: '2345678901' } });

    const addressInput = getAllByLabelText('Address *')[0];
    fireEvent.change(addressInput, { target: { value: '123 Main St' } });

    const firstNameParentInput = getAllByLabelText('First Name *')[1];
    fireEvent.change(firstNameParentInput, { target: { value: 'Jane' } });

    const lastNameParentInput = getAllByLabelText('Last Name *')[1];
    fireEvent.change(lastNameParentInput, { target: { value: 'Smith' } });

    const emailParentInput = getAllByLabelText('Email *')[1];
    fireEvent.change(emailParentInput, {
      target: { value: 'jane.smith@example.com' },
    });

    const phoneParentInput = getAllByLabelText('Phone *')[1];
    fireEvent.change(phoneParentInput, { target: { value: '2345678901' } });

    const addressParentInput = getAllByLabelText('Address *')[1];
    fireEvent.change(addressParentInput, { target: { value: '123 Main St' } });

    const submitButton = getByRole('button', { name: 'Submit' });
    await userEvent.click(submitButton);

    await waitFor(() => {
      expect(addStudent).toHaveBeenCalled();
    });
  });

  it('submits the form and calls editStudent for existing students', async () => {
    const editProps = {
      ...defaultProps,
      defaultValues: { ...defaultProps.defaultValues, documentId: '123' },
    };

    (editStudent as jest.Mock).mockResolvedValueOnce({
      data: { documentId: '123' },
    });

    const { getByRole, getAllByLabelText, getByText } = renderWithProviders(
      <StudentForm {...editProps} />,
    );

    const firstNameInput = getAllByLabelText('First Name *')[0];
    fireEvent.change(firstNameInput, { target: { value: 'David' } });

    const fileInput = getByText('Drag and drop or click here to select file');
    const file = new File(['content'], 'test-file.jpg', { type: 'image/jpeg' });
    fireEvent.change(fileInput, { target: { files: [file] } });

    const submitButton = getByRole('button', { name: 'Submit' });

    await userEvent.click(submitButton);

    await waitFor(() => {
      expect(editStudent).toHaveBeenCalled();
    });
  });
});
