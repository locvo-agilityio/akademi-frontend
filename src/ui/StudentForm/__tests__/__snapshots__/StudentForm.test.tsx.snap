// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentForm renders snapshot with default values 1`] = `
<div>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    class="chakra-stack css-135x7u7"
  >
    <div
      class="chakra-stack css-1fjucff"
    >
      <h3
        class="chakra-text css-c1j1q0"
      >
        Student Details
      </h3>
      <div
        class="chakra-stack css-r71kai"
      >
        <div
          class="chakra-stack css-enj3l8"
        >
          <label
            class="chakra-form__label css-11rkuxk"
          >
            Photo *
          </label>
          <div
            class="css-ptxt6o"
          >
            <p
              class="chakra-text css-144ed6u"
            >
              Drag and drop or click here to select file
            </p>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <input
                accept="image/*"
                class="chakra-input css-1k7fsj"
                hidden=""
                id="field-:r0:"
                title="Upload Image"
                type="file"
              />
            </div>
          </div>
        </div>
        <div
          class="css-idd5nh"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:r1:"
              id="field-:r1:-label"
            >
              First Name *
            </label>
            <input
              aria-label="first name"
              class="chakra-input css-172mybz"
              id="field-:r1:"
              name="firstName"
              placeholder="Please enter your first name"
              value="Johnny "
            />
          </div>
          <div
            class="css-kd4nig"
          >
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-11rkuxk"
                for="field-:r2:"
                id="field-:r2:-label"
              >
                Date & Place of Birth *
              </label>
              <input
                aria-label="birthday"
                class="chakra-input css-172mybz"
                id="field-:r2:"
                max="2009-12-31"
                name="birthday"
                placeholder="Please enter your birthday"
                type="date"
                value="1999-03-25"
              />
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <input
                aria-label="city"
                class="chakra-input css-172mybz"
                id="field-:r3:"
                name="city"
                placeholder="Please enter your city"
                value="Jakarta"
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:r4:"
              id="field-:r4:-label"
            >
              Email *
            </label>
            <input
              aria-label="email"
              class="chakra-input css-172mybz"
              id="field-:r4:"
              name="email"
              placeholder="Please enter your email"
              type="email"
              value="john.doe@example.com"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:r5:"
              id="field-:r5:-label"
            >
              Address
               *
            </label>
            <textarea
              aria-label="address"
              class="chakra-textarea css-1y43o19"
              id="field-:r5:"
              name="address"
              placeholder="Please enter your address"
            >
              Jakarta
            </textarea>
          </div>
        </div>
        <div
          class="css-idd5nh"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:r6:"
              id="field-:r6:-label"
            >
              Last Name *
            </label>
            <input
              aria-label="lastName"
              class="chakra-input css-172mybz"
              id="field-:r6:"
              name="lastName"
              placeholder="Please enter your last name"
              type="text"
              value="Ahmad"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:r7:"
              id="field-:r7:-label"
            >
              Parent Name *
            </label>
            <input
              aria-label="parentName"
              class="chakra-input css-172mybz"
              id="field-:r7:"
              name="parentName"
              placeholder="Please enter your parent name"
              type="text"
              value="John Doe"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:r8:"
              id="field-:r8:-label"
            >
              Phone *
            </label>
            <input
              aria-label="phone"
              class="chakra-input css-172mybz"
              id="field-:r8:"
              name="phone"
              placeholder="Please enter your phone"
              type="text"
              value="2345678901"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:r9:"
              id="field-:r9:-label"
            >
              Class *
            </label>
            <div
              class="chakra-radio-group css-uzxjvg"
              role="radiogroup"
            >
              <label
                class="chakra-radio css-1cqh9jq"
                data-checked=""
                for="radio-:rb:"
              >
                <input
                  checked=""
                  class="chakra-radio__input"
                  id="radio-:rb:"
                  name="grade"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="radio"
                  value="VII A"
                />
                <span
                  aria-hidden="true"
                  class="chakra-radio__control css-al5ct7"
                  data-checked=""
                />
                <span
                  class="chakra-radio__label css-vdbtus"
                  data-checked=""
                >
                  <p
                    class="chakra-text css-1427dgy"
                  >
                    VII A
                  </p>
                </span>
              </label>
              <label
                class="chakra-radio css-1cqh9jq"
                for="radio-:rc:"
              >
                <input
                  class="chakra-radio__input"
                  id="radio-:rc:"
                  name="grade"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="radio"
                  value="VII B"
                />
                <span
                  aria-hidden="true"
                  class="chakra-radio__control css-al5ct7"
                />
                <span
                  class="chakra-radio__label css-vdbtus"
                >
                  <p
                    class="chakra-text css-138ntzw"
                  >
                    VII B
                  </p>
                </span>
              </label>
              <label
                class="chakra-radio css-1cqh9jq"
                for="radio-:rd:"
              >
                <input
                  class="chakra-radio__input"
                  id="radio-:rd:"
                  name="grade"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="radio"
                  value="VII C"
                />
                <span
                  aria-hidden="true"
                  class="chakra-radio__control css-al5ct7"
                />
                <span
                  class="chakra-radio__label css-vdbtus"
                >
                  <p
                    class="chakra-text css-l85hjc"
                  >
                    VII C
                  </p>
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-stack css-1fjucff"
    >
      <h3
        class="chakra-text css-c1j1q0"
      >
        Parent Details
      </h3>
      <div
        class="chakra-stack css-r71kai"
      >
        <div
          class="css-idd5nh"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:re:"
              id="field-:re:-label"
            >
              First Name *
            </label>
            <input
              aria-label="firstName"
              class="chakra-input css-172mybz"
              id="field-:re:"
              name="parent.firstName"
              placeholder="Please enter your first name"
              type="text"
              value="John"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:rf:"
              id="field-:rf:-label"
            >
              Email *
            </label>
            <input
              aria-label="email"
              class="chakra-input css-172mybz"
              id="field-:rf:"
              name="parent.email"
              placeholder="Please enter your email"
              type="email"
              value="john.doe@example.com"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:rg:"
              id="field-:rg:-label"
            >
              Address
               *
            </label>
            <textarea
              aria-label="address"
              class="chakra-textarea css-1y43o19"
              id="field-:rg:"
              name="parent.address"
              placeholder="Please enter your address"
            >
              Jakarta
            </textarea>
          </div>
        </div>
        <div
          class="css-idd5nh"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:rh:"
              id="field-:rh:-label"
            >
              Last Name *
            </label>
            <input
              aria-label="lastName"
              class="chakra-input css-172mybz"
              id="field-:rh:"
              name="parent.lastName"
              placeholder="Please enter your last name"
              type="text"
              value="Doe"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:ri:"
              id="field-:ri:-label"
            >
              Phone *
            </label>
            <input
              aria-label="parentPhone"
              class="chakra-input css-172mybz"
              id="field-:ri:"
              name="parent.phone"
              placeholder="Please enter your phone"
              type="text"
              value="2345678901"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-11rkuxk"
              for="field-:rj:"
              id="field-:rj:-label"
            >
              Payments *
            </label>
            <div
              class="chakra-radio-group css-0"
              role="radiogroup"
            >
              <div
                class="chakra-stack css-1f0wxn3"
              >
                <label
                  class="chakra-radio css-1cqh9jq"
                  for="radio-:rl:"
                >
                  <input
                    class="chakra-radio__input"
                    id="radio-:rl:"
                    name="payment"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="cash"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-al5ct7"
                  />
                  <span
                    class="chakra-radio__label css-vdbtus"
                  >
                    Cash
                  </span>
                </label>
                <label
                  class="chakra-radio css-1cqh9jq"
                  for="radio-:rm:"
                >
                  <input
                    class="chakra-radio__input"
                    id="radio-:rm:"
                    name="payment"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="debit"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-al5ct7"
                  />
                  <span
                    class="chakra-radio__label css-vdbtus"
                  >
                    Debit
                  </span>
                </label>
                <label
                  class="chakra-radio css-1cqh9jq"
                  data-checked=""
                  for="radio-:rn:"
                >
                  <input
                    checked=""
                    class="chakra-radio__input"
                    id="radio-:rn:"
                    name="payment"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="unpaid"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-al5ct7"
                    data-checked=""
                  />
                  <span
                    class="chakra-radio__label css-vdbtus"
                    data-checked=""
                  >
                    Unpaid
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-stack css-tqtgzm"
    >
      <a
        class="chakra-button css-1o2131d"
        href="/students"
      >
        Cancel
      </a>
      <button
        class="chakra-button css-1o2131d"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
